In [1]:
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn import preprocessing
import seaborn as sns
import matplotlib.pyplot as plt
pd.set_option('display.max_columns',None)
%matplotlib notebook

In [2]:
train_df = pd.read_csv(Path('Resources/2019loans.csv'))
test_df = pd.read_csv(Path('Resources/2020Q1loans.csv'))

In [3]:
print(f" training set shape: {train_df.shape}")
print(f" testing set shape: {test_df.shape}")

 training set shape: (12180, 86)
 testing set shape: (4702, 86)


In [4]:
train_df.head()

Unnamed: 0   index  loan_amnt  int_rate  installment home_ownership  \
0       57107   57107    13375.0    0.1797       483.34       MORTGAGE   
1      141451  141451    21000.0    0.1308       478.68       MORTGAGE   
2      321143  321143    20000.0    0.1240       448.95       MORTGAGE   
3       11778   11778     3000.0    0.1240       100.22           RENT   
4      169382  169382    30000.0    0.1612      1056.49       MORTGAGE   

   annual_inc verification_status loan_status pymnt_plan    dti  delinq_2yrs  \
0    223000.0        Not Verified    low_risk          n  29.99          0.0   
1    123000.0     Source Verified    low_risk          n  11.26          2.0   
2    197000.0     Source Verified    low_risk          n  11.28          0.0   
3     45000.0        Not Verified    low_risk          n  18.08          0.0   
4    133000.0     Source Verified    low_risk          n  27.77          0.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  total_acc  \
0             0.0      15.0      0.0    39728.0       38.0   
1             0.0      16.0      0.0     9585.0       20.0   
2             0.0      12.0      0.0    16708.0       36.0   
3             0.0      12.0      1.0     8809.0       19.0   
4             2.0      13.0      0.0    65420.0       51.0   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   w   12804.66       12804.66       953.33           953.33   
1                   w   15152.97       15152.97     10037.02         10037.02   
2                   w   15724.23       15724.23      6971.52          6971.52   
3                   w    2930.77        2930.77        81.63            81.63   
4                   w   20778.03       20778.03     14763.99         14763.99   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0           570.34         382.99                 0.0         0.0   
1          5847.03        4189.99                 0.0         0.0   
2          4275.77        2695.75                 0.0         0.0   
3            69.23          12.40                 0.0         0.0   
4          9221.97        5542.02                 0.0         0.0   

   collection_recovery_fee  last_pymnt_amnt  collections_12_mths_ex_med  \
0                      0.0           483.34                         0.0   
1                      0.0           478.68                         0.0   
2                      0.0           448.95                         0.0   
3                      0.0            85.76                         1.0   
4                      0.0          1056.49                         0.0   

   policy_code application_type  acc_now_delinq  tot_coll_amt  tot_cur_bal  \
0          1.0       Individual             0.0           0.0     465790.0   
1          1.0       Individual             0.0           0.0      60769.0   
2          1.0       Individual             0.0           0.0     617946.0   
3          1.0       Individual             0.0         568.0      20761.0   
4          1.0       Individual             0.0           0.0     237839.0   

   open_acc_6m  open_act_il  open_il_12m  open_il_24m  mths_since_rcnt_il  \
0          2.0          7.0          3.0          7.0                 7.0   
1          0.0          2.0          0.0          1.0                17.0   
2          1.0          7.0          1.0          2.0                 5.0   
3          2.0          4.0          2.0          2.0                 8.0   
4          2.0          3.0          1.0          3.0                 5.0   

   total_bal_il  il_util  open_rv_12m  open_rv_24m  max_bal_bc  all_util  \
0       82290.0     70.0          1.0          1.0     24683.0      69.0   
1       18311.0     52.0          0.0          4.0      4311.0      50.0   
2       97335.0    108.0          0.0          0.0     11026.0      98.0   
3       11952.0     78.0          2.0          4.0      2409.0      49.0   
4       43636.0     74.0          1.0  

In [5]:
test_df.head()

Unnamed: 0  index  loan_amnt  int_rate  installment home_ownership  \
0       67991  67991    40000.0    0.0819       814.70       MORTGAGE   
1       25429  25429     6000.0    0.1524       208.70           RENT   
2       38496  38496     3600.0    0.1695       128.27           RENT   
3       19667  19667    20000.0    0.1524       478.33           RENT   
4       37505  37505     3600.0    0.1240       120.27           RENT   

   annual_inc verification_status loan_status pymnt_plan    dti  delinq_2yrs  \
0    140000.0        Not Verified    low_risk          n  19.75          0.0   
1     55000.0        Not Verified    low_risk          n  11.52          2.0   
2     42000.0        Not Verified    low_risk          n   6.74          0.0   
3    100000.0        Not Verified    low_risk          n  12.13          0.0   
4     50000.0        Not Verified    low_risk          n  16.08          0.0   

   inq_last_6mths  open_acc  pub_rec  revol_bal  total_acc  \
0             1.0      18.0      0.0     9471.0       45.0   
1             0.0       8.0      0.0     1280.0       12.0   
2             0.0       6.0      0.0     4757.0        6.0   
3             2.0       7.0      0.0    12731.0       21.0   
4             3.0       6.0      0.0    10413.0        6.0   

  initial_list_status  out_prncp  out_prncp_inv  total_pymnt  total_pymnt_inv  \
0                   w   33249.95       33249.95      9740.00          9740.00   
1                   w       0.00           0.00      6185.59          6185.59   
2                   w       0.00           0.00      3603.73          3603.73   
3                   w   17624.06       17624.06      4766.37          4766.37   
4                   w       0.00           0.00      3602.48          3602.48   

   total_rec_prncp  total_rec_int  total_rec_late_fee  recoveries  \
0          6750.05        2989.95                 0.0         0.0   
1          6000.00         185.59                 0.0         0.0   
2          3600.00           3.73                 0.0         0.0   
3          2375.94        2390.43                 0.0         0.0   
4          3600.00           2.48                 0.0         0.0   

   collection_recovery_fee  last_pymnt_amnt  collections_12_mths_ex_med  \
0                      0.0           814.70                         0.0   
1                      0.0          5773.27                         0.0   
2                      0.0          3607.12                         0.0   
3                      0.0           478.33                         0.0   
4                      0.0          3604.96                         0.0   

   policy_code application_type  acc_now_delinq  tot_coll_amt  tot_cur_bal  \
0          1.0       Individual             0.0           0.0     377773.0   
1          1.0       Individual             0.0           0.0      23460.0   
2          1.0       Individual             0.0           0.0      19183.0   
3          1.0       Individual             0.0           0.0      43817.0   
4          1.0       Individual             0.0           0.0      32448.0   

   open_acc_6m  open_act_il  open_il_12m  open_il_24m  mths_since_rcnt_il  \
0          1.0          6.0          1.0          2.0                 9.0   
1          1.0          1.0          1.0          1.0                 6.0   
2          0.0          3.0          0.0          0.0                40.0   
3          1.0          2.0          1.0          3.0                11.0   
4          2.0          2.0          1.0          1.0                 9.0   

   total_bal_il  il_util  open_rv_12m  open_rv_24m  max_bal_bc  all_util  \
0       61443.0     51.0          1.0          1.0      3218.0      26.0   
1       22180.0     94.0          0.0          1.0       713.0      68.0   
2       14426.0     96.0          1.0          2.0      4757.0      83.0   
3       31086.0     86.0          3.0          3.0      5917.0      78.0   
4       22035.0     88.0          2.0        

In [6]:
#categoriccal data headers in training set
train_cat_headers = train_df.columns[train_df.dtypes=="object"]
#number of unique values in each categorical feature
train_df[train_cat_headers].nunique()

home_ownership          4
verification_status     3
loan_status             2
pymnt_plan              1
initial_list_status     2
application_type        2
hardship_flag           2
debt_settlement_flag    2
dtype: int64

In [9]:
#repeat for testting set
test_cat_headers = test_df.columns[test_df.dtypes=="object"]
test_df[test_cat_headers].nunique()

home_ownership          4
verification_status     3
loan_status             2
pymnt_plan              1
initial_list_status     2
application_type        2
hardship_flag           2
debt_settlement_flag    1
dtype: int64

In [10]:
#create a dataframe to compare categorical data in training and testing set
compare_df = pd.DataFrame({"train_df":train_df[train_cat_headers].nunique(),
                          "test_df":test_df[test_cat_headers].nunique()})
print(train_df['debt_settlement_flag'].unique())
print(test_df['debt_settlement_flag'].unique())
compare_df

['N' 'Y']
['N']


train_df  test_df
home_ownership               4        4
verification_status          3        3
loan_status                  2        2
pymnt_plan                   1        1
initial_list_status          2        2
application_type             2        2
hardship_flag                2        2
debt_settlement_flag         2        1

### Observation
__debt_settlement_flag feature does not have 2 unique values in testing set, as a result after one hot encoding X_test will have one less column than X_train. The missing columns must be added to X_test__

In [46]:
# dropping index columns and loan_staus columns to prepare X_train and X_test
X_train = train_df.drop(['Unnamed: 0','index','loan_status'], axis=1)
X_train = pd.get_dummies(X_train, drop_first=True)
print(X_train.shape)

X_test = test_df.drop(['Unnamed: 0','index','loan_status','pymnt_plan'], axis=1)
X_test = pd.get_dummies(X_test, drop_first=True)
print(X_test.shape)


(12180, 85)
(4702, 84)


__Adding missing feature to X_test__

In [47]:
X_test['debt_settlement_flag_Y'] = 0
print(X_test.shape)

(4702, 85)


In [48]:
X_train.columns

Index(['loan_amnt', 'int_rate', 'installment', 'annual_inc', 'dti',
       'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal',
       'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_in

In [49]:
X_test.columns

Index(['loan_amnt', 'int_rate', 'installment', 'annual_inc', 'dti',
       'delinq_2yrs', 'inq_last_6mths', 'open_acc', 'pub_rec', 'revol_bal',
       'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_amnt', 'collections_12_mths_ex_med', 'policy_code',
       'acc_now_delinq', 'tot_coll_amt', 'tot_cur_bal', 'open_acc_6m',
       'open_act_il', 'open_il_12m', 'open_il_24m', 'mths_since_rcnt_il',
       'total_bal_il', 'il_util', 'open_rv_12m', 'open_rv_24m', 'max_bal_bc',
       'all_util', 'total_rev_hi_lim', 'inq_fi', 'total_cu_tl', 'inq_last_12m',
       'acc_open_past_24mths', 'avg_cur_bal', 'bc_open_to_buy', 'bc_util',
       'chargeoff_within_12_mths', 'delinq_amnt', 'mo_sin_old_il_acct',
       'mo_sin_old_rev_tl_op', 'mo_sin_rcnt_rev_tl_op', 'mo_sin_rcnt_tl',
       'mort_acc', 'mths_since_recent_bc', 'mths_since_recent_in

__Check if all X_train columns are present in X_test__. 


In [35]:
np.setdiff1d(X_train.columns, X_test.columns)

array([], dtype=object)

__Empty array indicates all X_train columns are present in X_test__

In [50]:
le = preprocessing.LabelEncoder()
y_train = le.fit_transform(train_df["loan_status"])
print(le.classes_)
print(y_train)
y_test = le.fit_transform(test_df["loan_status"])
print(y_test)
test_df["loan_status"].value_counts()

['high_risk' 'low_risk']
[1 1 1 ... 0 0 0]
[1 1 1 ... 0 0 0]


high_risk    2351
low_risk     2351
Name: loan_status, dtype: int64

### Logistic Regression vs Random Forest Classifier
Random Forest Classifier is expected to perform better than Logistic Regression as it does not assume a linear relationship between the label and the features. Random Forest Classifier creates many decision trees and combines the final result, as the sample size grows larger Random Forest Classifier has better accuracy in prediction 

In [51]:
from sklearn.linear_model import LogisticRegression
model_lr = LogisticRegression()
model_lr.fit(X_train, y_train)
print("LR Train Score: ",model_lr.score(X_train,y_train))
print("LR Test Score: ",model_lr.score(X_test,y_test))

LR Train Score:  0.6497536945812807
LR Test Score:  0.5157379838366652


/Users/mohanpanga/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [52]:
from sklearn.ensemble import RandomForestClassifier
model_rfc = RandomForestClassifier(n_estimators=15, random_state = 0)
model_rfc.fit(X_train, y_train)
print("RFC Train Score: ",model_rfc.score(X_train,y_train))
print("RFC Test Score: ",model_rfc.score(X_test,y_test))

RFC Train Score:  0.99688013136289
RFC Test Score:  0.6101658868566567


__Random Forest Classifer has better score than Logistic Regression__

In [ ]:
# scores_train = []
# scores_test = []
# for i in range(1,50):
#     clf = RandomForestClassifier(n_estimators=i, random_state=0)
#     clf.fit(X_train,y_train)
#     scores_train.append(clf.score(X_train, y_train))
#     scores_test.append(clf.score(X_test,y_test))
#     print(clf.score(X_train, y_train))
    
# #     clf.predict(X_test)
# x=range(1,50)
# sns.scatterplot(x,scores_train)
# plt.show()

### Preprocessing data
Scaling the data should typically improve the accuracy of the model as large values such as loan amount and income do not overly influence the model predictions

In [58]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler().fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)
X_test_scaled

array([[ 2.20755943, -1.12001617,  1.0371484 , ..., -0.41370744,
        -0.17149859, -0.02026518],
       [-1.11348584,  0.21833096, -1.01983876, ..., -0.41370744,
        -0.17149859, -0.02026518],
       [-1.34791257,  0.54295132, -1.2928478 , ..., -0.41370744,
        -0.17149859, -0.02026518],
       ...,
       [-0.72277464,  1.7009538 , -0.41340093, ..., -0.41370744,
        -0.17149859, -0.02026518],
       [-0.91813024,  0.85997823, -1.02947877, ..., -0.41370744,
        -0.17149859, -0.02026518],
       [ 1.23078141,  1.22636262,  2.08478621, ..., -0.41370744,
        -0.17149859, -0.02026518]])

In [62]:
model_lr_scaled = LogisticRegression()
model_lr_scaled.fit(X_train_scaled, y_train)
print("LR_scaled Train Score: ",model_lr_scaled.score(X_train_scaled,y_train))
print("LR_scaled Test Score: ",model_lr_scaled.score(X_test_scaled,y_test))

model_rfc_scaled = RandomForestClassifier(n_estimators=15, random_state = 0)
model_rfc_scaled.fit(X_train_scaled, y_train)
print("RFC_scaled Train Score: ",model_rfc_scaled.score(X_train_scaled,y_train))
print("RFC_scaled Test Score: ",model_rfc_scaled.score(X_test_scaled,y_test))

/Users/mohanpanga/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LR_scaled Train Score:  0.7083743842364532
LR_scaled Test Score:  0.7681837515950659
RFC_scaled Train Score:  0.9967159277504105
RFC_scaled Test Score:  0.6091025095703956


__Scaling the data improved the logistic regression model predictions. The train and test  scores increased from 0.649, 0.515 to 0.708, 0.768 respectively. There is significant improvement in prediction. It is likely that the large numbers in some of the features are affecting the prediction. Scaling the data helps in giving an equal weight to all features.__

__RCF model predictions did not change much with scaling. It is likely because the decision tree technique does not assume a linear relationship and does not have model coefficients that can be affected the magnitude of the values in the features__


In [63]:
# Convert categorical data to numeric and separate target feature for training data


In [ ]:
# Convert categorical data to numeric and separate target feature for testing data

In [ ]:
# add missing dummy variables to testing set

In [ ]:
# Train the Logistic Regression model on the unscaled data and print the model score

In [ ]:
# Train a Random Forest Classifier model and print the model score

In [ ]:
# Scale the data

In [ ]:
# Train the Logistic Regression model on the scaled data and print the model score

In [ ]:
# Train a Random Forest Classifier model on the scaled data and print the model score